bitcoin/src/blockdata/transaction.rs

```diff
@@ -1148,6 +1170,115 @@ impl From<&Transaction> for Wtxid {
     }
 }
 
+/// Predicts the weight of a to-be-constructed transaction.
+///
+/// This function computes the weight of a transaction which is not fully known. All that is needed
+/// is the lengths of scripts and witness elements.
+///
+/// # Arguments
+///
+/// * `inputs` - an iterator which returns `InputWeightPrediction` for each input of the
+///   to-be-constructed transaction.
+/// * `output_script_lens` - an iterator which returns the length of `script_pubkey` of each output
+///   of the to-be-constructed transaction.
+///
+/// Note that lengths of the scripts and witness elements must be non-serialized, IOW *without* the
+/// preceding compact size. The lenght of preceding compact size is computed and added inside the
+/// function for convenience.
+///
+/// If you  have the transaction already constructed (except for signatures) with a dummy value for
+/// fee output you can use the return value of [`Transaction::script_pubkey_lens`] method directly
+/// as the second argument.
+///
+/// # Usage
+///
+/// When signing a transaction one doesn't know the signature before knowing the transaction fee and
+/// the transaction fee is not known before knowing the transaction size which is not known before
+/// knowing the signature. This apparent dependency cycle can be broken by knowing the length of the
+/// signature without knowing the contents of the signature e.g., we know all Schnorr signatures
+/// are 64 bytes long.
```

Comment:
Hey, I just saw this PR mentioned in the Bitcoin Optech Newsletter, and wanted to point out that there is a trick to avoiding the catch-22 between transaction weight, input count, and fees. When you build a transaction, you know the recipient outputs you wish to create. You also know the size of the metadata (version, input counter, output counter, locktime, optionally witness marker and witness flag). Before building a transaction, you usually also know the feerate that you wish to achieve with the transaction.

This allows you to calculate the transaction weight of the “fixed part” of the transaction, and per the feerate, the necessary fees for that fixed part. For the inputs, you calculate their _effective value_. For each input you calculate the weight it will take to add the specific input to a transaction, multiply that weight with the feerate, and deduct this “input fee” from that input’s value. After this preprocessing step, when you pick UTXOs to fund the transaction, you count only their _effective value_ towards the target (i.e. recipient amounts plus “fixed parts fees”)—this makes the selection of inputs fee-neutral because the effective value accounts for the input paying for itself. The target can additionally be adjusted for whether you expect to create a change output, how valuable the change should be at least, and what output type the change is intended to have.

I described an efficient transaction building process using UTXO’s effective value here on [Stack Exchange](https://bitcoin.stackexchange.com/a/98394/5406).

Reply:
the very last bullet https://bitcoinops.org/en/newsletters/2023/02/22/

Reply:
@Kixunil: Sorry, I was looking at the open PR before the Newsletter got published this morning. :)